bridge

esco opened the pull request.

philipstanislaus commented on bridge/bridge.go:
```diff
@@ -0,0 +1,84 @@
+	fmt.Printf("account: %+v\n", acc)
+	// jsonValue, err := b.cdc.MarshalJSON(acctBytes)
+	if err != nil {
+		panic(err)
```
Does that lead to a runtime error on the JavaScript side? Or does it completely crash the WASM module?

philipstanislaus commented on bridge/bridge.go:
```diff
@@ -0,0 +1,84 @@
+	respBytes := json.RawMessage(strValue)
+	fmt.Printf("strValue: %+v\n, respBytes: %+v\n", strValue, respBytes)
+	resp := &types.ResultABCIQuery{}
+	util.UnmarshalResponseBytes(b.cdc, respBytes, resp)
```
Should what handle this error?

philipstanislaus commented on bridge/bridge.go:
```diff
@@ -0,0 +1,84 @@
+	coins := args[2].String()
+	privKeyHex := args[3].String()
+	txBytes, _ := tx.NewSendTx(from, to, coins, privKeyHex, b.cdc)
+	jsonValue, _ := b.cdc.MarshalJSON(txBytes)
```
Do we need to handle this error?

philipstanislaus commented on bridge/bridge.go:
```diff
@@ -0,0 +1,84 @@
+	to := args[1].String()
+	coins := args[2].String()
+	privKeyHex := args[3].String()
+	txBytes, _ := tx.NewSendTx(from, to, coins, privKeyHex, b.cdc)
```
Should we not handle this error?

musnit commented:
great comments - i'll look in more detail after devcon too

On Sun, 28 Oct 2018 at 16:40 philipstanislaus <notifications@github.com>
wrote:

> *@philipstanislaus* commented on this pull request.
>
> Great work @esco <https://github.com/esco> 💯, it's exciting to see this
> works!
>
> A few notes:
>
>    1. The interaction architecture between go and JavaScript is hard to
>    understand. I think this is due to 2 things: a) the back and forth of
>    callbacks and b) a lack of abstraction between the lower communication
>    layer (set up of the bridge, creating the client, managing namespace and
>    function names, callbacks) and the higher application layer (registering
>    the functions we actually want to use). Unfortunately, I think we do not
>    have time to refactor the functionality. I also understand that the matter
>    is rather complex. For now, it might help to have a short description or a
>    link to understand the back and forth between Go and JavaScript on bridge
>    setup (from Go calling JavaScript to get namespace, from JavaScript calling
>    Go to request functions etc...).
>    2. Independent of the architecture, I strongly suggest that we have a
>    few end to end tests (creating the bridge, calling go functions, verifying
>    results).
>    3. Since this project will receive attention from Tendermint and other
>    Cosmos based teams, we should have a quick linting setup, I'd suggest to
>    use the same configs we already use in other THORChain repos:
>    https://standardjs.com/ for JavaScript and this config:
>    https://github.com/thorchain/THORChain/blob/master/tools/gometalinter.json
>    for Go.
>    4. For the same reason as 3, I added Code Climate, please review
>    issues:
>    https://codeclimate.com/github/thorchain/thorchain-wasm-bridge/pull/1
>
> ------------------------------
>
> In bridge/bridge.go
> <https://github.com/thorchain/thorchain-wasm-bridge/pull/1#discussion_r228752435>
> :
>
> > +		callback.Release()
> +	}
> +}
> +
> +func (b *Bridge) decodeAccount(args []js.Value) {
> +	strValue := args[0].String()
> +	respBytes := json.RawMessage(strValue)
> +	fmt.Printf("strValue: %+v\n, respBytes: %+v\n", strValue, respBytes)
> +	resp := &types.ResultABCIQuery{}
> +	util.UnmarshalResponseBytes(b.cdc, respBytes, resp)
> +	fmt.Printf("resp.Value: %+v\n", resp.Response.Value)
> +	acc, err := util.DecodeAccount(b.cdc, resp.Response.Value)
> +	fmt.Printf("account: %+v\n", acc)
> +	// jsonValue, err := b.cdc.MarshalJSON(acctBytes)
> +	if err != nil {
> +		panic(err)
>
> Does that lead to a runtime error on the JavaScript side? Or does it
> completely crash the WASM module?
> ------------------------------
>
> In bridge/bridge.go
> <https://github.com/thorchain/thorchain-wasm-bridge/pull/1#discussion_r228752478>
> :
>
> > +		callbackDecodeAcct,
> +	}
> +}
> +
> +func (b *Bridge) ReleaseCallbacks() {
> +	for _, callback := range b.callbacks {
> +		callback.Release()
> +	}
> +}
> +
> +func (b *Bridge) decodeAccount(args []js.Value) {
> +	strValue := args[0].String()
> +	respBytes := json.RawMessage(strValue)
> +	fmt.Printf("strValue: %+v\n, respBytes: %+v\n", strValue, respBytes)
> +	resp := &types.ResultABCIQuery{}
> +	util.UnmarshalResponseBytes(b.cdc, respBytes, resp)
>
> Should what handle this error?
> ------------------------------
>
> In bridge/bridge.go
> <https://github.com/thorchain/thorchain-wasm-bridge/pull/1#discussion_r228752520>
> :
>
> > +	}
> +	jsonValue, err := b.cdc.MarshalJSON(acc)
> +	fmt.Print("jsonValue: +%v\n", jsonValue)
> +	callback := args[len(args)-1:][0]
> +	jsCallback := js.Global().Get(callback.String())
> +	jsCallback.Invoke(string(jsonValue))
> +}
> +
> +func (b *Bridge) sendMessage(args []js.Value) {
> +	fmt.Printf("args: %+v", args)
> +	from := args[0].String()
> +	to := args[1].String()
> +	coins := args[2].String()
> +	privKeyHex := args[3].String()
> +	txBytes, _ := tx.NewSendTx(from, to, coins, privKeyHex, b.cdc)
> +	jsonValue, _ := b.cdc.MarshalJSON(txBytes)
>
> Do we need to handle this error?
> ------------------------------
>
> In bridge/bridge.go
> <https://github.com/thorchain/thorchain-wasm-bridge/pull/1#discussion_r228752537>
> :
>
> > +		panic(err)
> +	}
> +	jsonValue, err := b.cdc.MarshalJSON(acc)
> +	fmt.Print("jsonValue: +%v\n", jsonValue)
> +	callback := args[len(args)-1:][0]
> +	jsCallback := js.Global().Get(callback.String())
> +	jsCallback.Invoke(string(jsonValue))
> +}
> +
> +func (b *Bridge) sendMessage(args []js.Value) {
> +	fmt.Printf("args: %+v", args)
> +	from := args[0].String()
> +	to := args[1].String()
> +	coins := args[2].String()
> +	privKeyHex := args[3].String()
> +	txBytes, _ := tx.NewSendTx(from, to, coins, privKeyHex, b.cdc)
>
> Should we not handle this error?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/thorchain/thorchain-wasm-bridge/pull/1#pullrequestreview-169096322>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAtZufpBtmz1fNTxIxVtS_cL1e8Qf93cks5updABgaJpZM4X4bll>
> .
>

esco commented:
Thanks for the feedback @philipstanislaus 

> The interaction architecture between go and JavaScript is hard to understand. I think this is due to 2 things: a) the back and forth of callbacks

I agree the callback based communication is cumbersome. You can't get a response from a Go call into Javascript by default. I abstracted away a lot of the nuance of getting the response of a function for the end user by wrapping all callbacks with a promise that will be resolved with the result of the Go function. Given only the name of a Go function that you want to export, `requestCallback` will handle the back and forth communication necessary.

Example Usage:
```
const account = await client.getAccount(address) // <-- the response comes from the wrapper
```

> and b) a lack of abstraction between the lower communication layer (set up of the bridge, creating the client, managing namespace and function names, callbacks) and the higher application layer (registering the functions we actually want to use). 

The setup there is actually the abstraction layer (unless I misunderstand your point about lack of abstraction). The end user only has to pass in the location of the binary and uri of a node and they are only exposed to `runner` which has functions for loading the binary and executing it and `client` which has the Go methods. Example usage can be seen in `index.html` (usage will be a bit different in my next commit since I migrated to commonjs). In my update I'll add a usage section to the readme.

The namespace and function names shouldn't be managed by the end user. These are there to abstract the complexity of invoking and getting a response from a Go callback. I think one of your concerns might be the ability to specify specific functions from the binary. Thats a good point, currently the bridge is coupled to the binary. I could expose the `requestCallback` function to the end user so that the JS is more flexible across different binaries. This means that the `Client` class would need to be extended by the end user or they could access client.bridge directly.

> Unfortunately, I think we do not have time to refactor the functionality. I also understand that the matter is rather complex. For now, it might help to have a short description or a link to understand the back and forth between Go and JavaScript on bridge setup (from Go calling JavaScript to get namespace, from JavaScript calling Go to request functions etc...).

Agreed on documenting the Go/JS communication. If I understand the concern correctly I think the only refactor would be to expose `requestCallback`.

To make sure we're on the same page, this is the current end user api:

```
const { client, runner } = THORChainBridge('binary.wasm', 'http://localhost:26657/')
// Load binary
await runner.loadBinary()
// Execute
await runner.run()
console.log('bridge ready')
```

> Independent of the architecture, I strongly suggest that we have a few end to end tests (creating the bridge, calling go functions, verifying results).

Agreed

> Since this project will receive attention from Tendermint and other Cosmos based teams, we should have a quick linting setup, I'd suggest to use the same configs we already use in other THORChain repos: https://standardjs.com/ for JavaScript and this config: https://github.com/thorchain/THORChain/blob/master/tools/gometalinter.json for Go.
For the same reason as 3, I added Code Climate, please review issues: https://codeclimate.com/github/thorchain/thorchain-wasm-bridge/pull/1

Great point. I ported the code to TypeScript, added our tslint and made it a commonjs module. That coupled with some ideas about making `requestCallback` public and extending Client, would make the `bridge.js` part generic to any binary.

philipstanislaus commented:
@esco thanks for your comments.

Regarding a) callback based communication: I liked the way you used promises to abstract the callback hell away. What might help is a naming convention, like `responseCb` from the outside, and `done` within the function that calls it (we could probably come up with something better). But I'd leave it as is for time's sake now.

Regarding b) lack of abstraction: My wording was not concice enough I think, should have probably written: lack of visible abstraction. Calling functions as an end user looks great, but registering new methods in the go codebase and the js bridge needs to be done in the same files that handle the messaging. I would suggest to separate the communication and application logic into maybe 2 files, so it is easier to grasp where to add what. Decoupling the bridge from the binary is a good idea. Again – I think these changes are not necessary now, let's just keep them documented here for future reference :) 

Thanks for porting to TS and adding linting 🚀 💯

musnit commented:
@philipstanislaus @Jessicawatsonmiller 
I think as long as this is basically working, it's fine to leave for now and we should consider it to be a functional prototype.

In later milestones we should look into fixing the issues and refactorings phillip suggests, as well as do another security/performance review for further optimizations once we have a fully working MVP.

The whole go -> webassembly strategy for dealing with amino here is itself a short-term hack that, though may become long term sustainable, may be worth considering replacing with a full JS implementation later, so I would avoid premature optimization/investing time into this aslong we have it at the point where it can be used for everyday users to test and interact with asgardex.

(@Jessicawatsonmiller note: a few of these core optimization/security review action items are coming up across our different codebases. we should probably track them in a single place to be dealt with in a few months once super-security becomes a priority)
